pkg/webhook/resources/setting/validator.go

```diff
@@ -709,6 +730,201 @@ func (v *settingValidator) validateDeleteStorageNetwork(setting *v1beta1.Setting
 	return werror.NewMethodNotAllowed(fmt.Sprintf("Disallow delete setting name %s", settings.StorageNetworkName))
 }
 
+func (v *settingValidator) getPrometheusVolumes() ([]string, error) {
+	prometheus, err := v.prometheusCache.Get(util.CattleMonitoringSystemNamespace, util.RancherMonitoringPrometheus)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logrus.Infof("prometheus not found. skip")
+			return nil, nil
+		}
+		return nil, fmt.Errorf("prometheus get error %w", err)
+	}
+
+	if prometheus.Spec.Replicas == nil {
+		return nil, nil
+	}
+
+	prometheusVolumes := make([]string, 0)
+	for i := 0; i < int(*prometheus.Spec.Replicas); i++ {
+		pvc, err := v.pvcCache.Get(prometheus.Namespace, util.RancherMonitoringPrometheusPVCPrefix+strconv.Itoa(i))
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				logrus.Infof("prometheus pvc %v not found. skip", util.RancherMonitoringPrometheusPVCPrefix+strconv.Itoa(i))
+				continue
+			}
+			return nil, fmt.Errorf("prometheus get pvc error %w", err)
+		}
+
+		if pvc.Spec.VolumeName == "" {
+			continue
+		}
+
+		prometheusVolumes = append(prometheusVolumes, pvc.Spec.VolumeName)
+	}
+
+	return prometheusVolumes, nil
+}
+
+func (v *settingValidator) getAlertManagerVolumes() ([]string, error) {
+	alertmanager, err := v.alertmanagerCache.Get(util.CattleMonitoringSystemNamespace, util.RancherMonitoringAlertmanager)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logrus.Infof("alertmanager not found. skip")
+			return nil, nil
+		}
+		return nil, fmt.Errorf("alertmanager get error %w", err)
+	}
+
+	if alertmanager.Spec.Replicas == nil {
+		return nil, nil
+	}
+
+	alertmanagerVolumes := make([]string, 0)
+	for i := 0; i < int(*alertmanager.Spec.Replicas); i++ {
```

Comment:
Can you see if we have a better way to find alert manager volume? maybe via owner or labels. Thanks.

Reply:
As @yaocw2020 suggested https://github.com/harvester/harvester/pull/4124#discussion_r1250203133,
I think we could try to list PVCs with selector.

Reply:
https://github.com/WebberHuang1118/harvester/blob/121f387ad0ce6c83bfde536aeabd02bc6a02663b/pkg/webhook/resources/setting/validator.go#L748
Listing PVCs used by AlertManager via labels and selectors, thanks.